crates/kubelet/src/image_client.rs

```diff
@@ -10,33 +10,63 @@ use oci_distribution::Reference;
 /// An image client capable of fetching images from a storage location
 #[async_trait]
 pub trait ImageClient {
-    /// Given a certain image reference pull the image data from a storage location
+    /// Given a certain image reference pull the image data from a storage location.
+    ///
+    /// The default implementation pulls the image data and digest, and returns
+    /// the data.
+    async fn pull(&mut self, image_ref: &Reference) -> anyhow::Result<Vec<u8>> {
+        let (content, _) = self.pull_with_digest(image_ref).await?;
+        Ok(content)
+    }
+
+    /// Fetch the image data and, if available, image digest for the given image
+    /// reference from a storage location.
     ///
     /// # Example
     /// ```rust
     /// use async_trait::async_trait;
     /// use kubelet::image_client::ImageClient;
     /// use oci_distribution::Reference;
     ///
-    /// struct InMemoryClient(std::collections::HashMap<Reference, Vec<u8>>);
+    /// struct InMemoryClient(std::collections::HashMap<Reference, (Vec<u8>, String)>);
     ///
     /// #[async_trait]
     /// impl ImageClient for InMemoryClient {
-    ///     async fn pull(&mut self, image: &Reference) -> anyhow::Result<Vec<u8>> {
-    ///         let image = self
+    ///     async fn pull_with_digest(&mut self, image_ref: &Reference) -> anyhow::Result<(Vec<u8>, Option<String>)> {
+    ///         let (image, digest) = self
     ///             .0
-    ///             .get(image)
+    ///             .get(image_ref)
     ///             .ok_or(anyhow::anyhow!("Couldn't find image"))?;
-    ///         Ok(image.clone())
+    ///         Ok((image.clone(), Some(digest.to_string())))
     ///     }
     /// }
     /// ```
-    async fn pull(&mut self, image: &Reference) -> anyhow::Result<Vec<u8>>;
+    async fn pull_with_digest(
```

Comment:
You are probably right that we can just have `pull` return the content plus digest.  I did this because I didn't know how widely the existing `pull` was used.